src/main/java/com/visenze/productsearch/ProductSearch.java

```diff
@@ -0,0 +1,213 @@
+         * Build the ProductSearch class based on configurations set through the
+         * builder pattern.
+         *
+         * @return ProductSearch object class
+         */
+        public ProductSearch build() {
+            if (this.appKey == null || this.appKey.isEmpty()) {
+                throw new InternalViSearchException(ResponseMessages.INVALID_KEY);
+            }
+            return new ProductSearch(this.appKey, this.placementId, endpoint, config);
+        }
+
+        /**
+         * Set api custom url endpoint for
+         *
+         * @param url The custom url endpoint to use instead
+         *
+         * @return this 'itself'
+         */
+        public Builder setApiEndPoint(String url) {
+            this.endpoint = url;
+            return this;
+        }
+
+        /**
+         * Set custom client configurations for http connections
+         *
+         * @param config custom client http configurations
+         *
+         * @return this 'itself'
+         *
+         * @see ClientConfig
+         */
+        public Builder setClientConfig(ClientConfig config) {
+            this.config = config;
+            return this;
+        }
+    }
+
+    /**
+     * Constructor of ProductSearch made private to only allow it's Builder
+     * class to create it.
+     *
+     * @param appKey Unique app key that acts as authenticator for the client
+     * @param placementId Placement ID of the template that was chosen when
+     *                     creating apps on the dashboard
+     * @param endpoint Which endpoint to use for all ViHttpClient queries
+     * @param config Configuration to the behaviours of the ViHttpClient
+     */
+    private ProductSearch(String appKey, Integer placementId, String endpoint,
+                          ClientConfig config)
+    {
+        this.appKey       = appKey;
+        this.placementId  = placementId;
+        this.endpoint     = endpoint;
+        this.httpClient   = new ViSearchHttpClientImpl(this.endpoint, appKey, placementId.toString(), config);
+    }
+
+    /**
+     * Calls the POST method for the image search API
+     *
+     * @param params ImageSearchParam specific parameters to perform request
+     *
+     * @return ViSearchHttpResponse http response of search results
+     */
+    public ProductSearchResponse imageSearch(SearchByImageParam params) {
+        // test for image parameter validity
+        final File imageFile = params.getImage();
+        // attempt search using image file
+        if (imageFile != null) {
+            try {
+                return ProductSearchResponse.fromResponse(httpClient.postImage(DEFAULT_IMAGE_SEARCH_PATH, params.toMultimap(), new FileInputStream(imageFile), imageFile.getName()));
+            } catch (FileNotFoundException e) {
+                throw new InternalViSearchException(ResponseMessages.INVALID_IMAGE_OR_URL, e);
+            }
+        }
+        Multimap<String, String> paramMap = params.toMultimap();
```

Comment:
move these 3 lines to right at the beginning of the method (line 176). need to be in a common function that can be called by both imageSearch and visualSimilarSearch

Reply:
currently if imageFile is not null, this will never reach.